File: 一些小demo/blockDemo/blockDemo/ViewController.m
//
//  ViewController.m
//  blockDemo
//
//  Created by Changjun Li on 2024/9/28.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    __block int count = 10;
    // 定义一个栈上的 Block
    void (^myBlock)(void) = ^{
        // Block 内访问和修改 __block 变量
        NSLog(@"Inside Block - Before: %d", count);
        count += 5;
        NSLog(@"Inside Block - After: %d", count);
    };
    
    // 执行栈上的 Block
    NSLog(@"Before Block execution - count: %d", count);
    myBlock();
    NSLog(@"After Block execution - count: %d", count);
    
    // 将 Block 复制到堆上
    void (^heapBlock)(void) = [myBlock copy];
    
    // 执行堆上的 Block
    NSLog(@"Before Heap Block execution - count: %d", count);
    heapBlock();
    NSLog(@"After Heap Block execution - count: %d", count);
}


@end

File: 单元练习Demo/NSOperationDemo/NSOperationDemo/ViewController.swift
//
//  ViewController.swift
//  NSOperationDemo
//
//  Created by Changjun Li on 2024/10/1.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        // 创建一个 NSOperationQueue
        let operationQueue = OperationQueue()

        // 使用 NSBlockOperation 创建任务
        let operation1 = BlockOperation {
            for i in 1...3 {
                print("Operation 1: \(i)")
                sleep(1) // 模拟耗时任务
            }
        }
        let operation2 = BlockOperation {
            for i in 1...3 {
                print("Operation 2: \(i)")
                sleep(1) // 模拟耗时任务
            }
        }

        // 添加依赖，确保 operation2 在 operation1 完成后执行
        operation1.addDependency(operation2)

        // 将操作添加到队列中
        operationQueue.addOperation(operation1)
        operationQueue.addOperation(operation2)
        
        // 取消任务1
        operation1.cancel()
    }
}

